ledger/acctupdates.go

```diff
@@ -1473,6 +1482,86 @@ func (au *accountUpdates) upgradeDatabaseSchema3(ctx context.Context, tx *sql.Tx
 	return 4, nil
 }
 
+
+// upgradeDatabaseSchema4 upgrades the database schema from version 4 to version 5,
+// cleaning old empty catchpoint directories.
+func (au *accountUpdates) upgradeDatabaseSchema4(ctx context.Context, tx *sql.Tx) (updatedDBVersion int32, err error) {
+	//validate no empty dirs
+	err = au.removeEmptyDirsOnSchemaUpgrade()
+	if err != nil {
+		return 0, err
+	}
+
+	// update version
+	_, err = db.SetUserVersion(ctx, tx, 5)
+	if err != nil {
+		return 0, fmt.Errorf("accountsInitialize unable to update database schema version from 4 to 5: %v", err)
+	}
+	return 5, nil
+}
+
+
+// IsDirectoryEmpty returns if a given directory is empty or not.
+func IsDirectoryEmpty(path string) (bool, error){
+	dir, err := os.Open(path)
+	if err != nil{
+		return false, err
+	}
+	defer dir.Close()
+	_,err = dir.Readdirnames(1)
+	if err != io.EOF{
+		return false,err
+	}
+	return true,nil
+}
+
+func  (au *accountUpdates) removeEmptyDirsOnSchemaUpgrade() (err error) {
+	catchpointRootDir := filepath.Join(au.dbDirectory,CatchpointDirName)
+	if _,err := os.Stat(catchpointRootDir); os.IsNotExist(err) {
+		return nil
+	}
+	for {
+		emptyDirs, err := GetEmptyDirs(catchpointRootDir)
+		if err != nil {
+			return err
+		}
+		// There are no empty dirs
+		if len(emptyDirs) == 0 {
+			break
+		}
+		// only left with the root dir
+		if len(emptyDirs) == 1 && emptyDirs[0] == catchpointRootDir {
+			break
+		}
+		for _, emptyDirPath := range emptyDirs {
+			os.Remove(emptyDirPath)
+		}
+	}
+	return nil
+}
+
+// GetEmptyDirs returns a slice of paths for empty directories which are located in PathToScan arg
+func GetEmptyDirs(PathToScan string) ([]string, error) {
+	var emptyDir []string
+	err := filepath.Walk(PathToScan,func(path string, f os.FileInfo, errIn error) error {
+		if errIn != nil  {
+			return errIn
+		}
+		if !f.IsDir() {
+			return nil
+		}
+		isEmpty, err := IsDirectoryEmpty(path)
+		if err != nil {
```

Comment:
in case you get a file-doesn't-exist here, you can just ignore it and keep going. ( it just mean that it was deleted in the background by some other process. )